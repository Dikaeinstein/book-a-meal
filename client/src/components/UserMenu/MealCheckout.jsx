import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import history from '../../helpers/history';
import Button from '../util/Button';
import { checkoutOrder } from '../../actions/orderActions';

class ConnectedMealCheckout extends Component {
  constructor(props) {
    super(props);
    this.state = {
      total: parseFloat(props.meal.price),
      quantity: 1,
      amount: parseFloat(props.meal.price),
    };
    this.handleChange = this.handleChange.bind(this);
    this.handleCheckout = this.handleCheckout.bind(this);
  }

  handleChange(event) {
    event.preventDefault();
    const value = event.target.value;
    if (value && value >= 1) {
      const newQuantity = parseInt(value, 10);
      this.setState({
        quantity: newQuantity,
        total: newQuantity * this.state.amount,
      });
    }
  }

  handleCheckout() {
    this.props.closeModal();
    this.props.checkoutOrder({
      mealId: this.props.meal.id,
      name: this.props.meal.name,
      quantity: this.state.quantity,
      total: this.state.total,
      amount: this.props.meal.price,
    });
    history.push('/order-confirmation');
  }

  render() {
    return (
      <div id="checkout" className="text-center">
        <h3>Checkout Order</h3>
        <div>
          <table>
            <thead className="text-center">
              <tr>
                <th>
                  Meal Name
                </th>
                <th>
                  Quantity (Plate)
                </th>
                <th>
                  Total
                </th>
              </tr>
            </thead>
            <tbody>
              <tr className="text-center">
                <td>
                  {this.props.meal.name}
                </td>
                <td>
                  <input
                    type="number"
                    name="quantity"
                    id="quantity"
                    value={this.state.quantity}
                    onChange={this.handleChange}
                    min={1}
                    max={Number.MAX_SAFE_INTEGER}
                    style={{ width: '50%' }}
                  />
                </td>
                <td>
                  &#x20a6; {this.state.total}
                </td>
              </tr>
            </tbody>
          </table>
        </div>
        <Button
          className="btn btn-default"
          value="Checkout"
          style={{ margin: '1rem auto' }}
          onClick={this.handleCheckout}
        />
      </div>
    );
  }
}

ConnectedMealCheckout.propTypes = {
  meal: PropTypes.objectOf(PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.number,
  ])).isRequired,
  closeModal: PropTypes.func.isRequired,
  checkoutOrder: PropTypes.func.isRequired,
};

const mapDispatchToProps = dispatch => ({
  checkoutOrder(order) { dispatch(checkoutOrder(order)); },
});

const MealCheckout = connect(null, mapDispatchToProps)(ConnectedMealCheckout);

export default MealCheckout;
