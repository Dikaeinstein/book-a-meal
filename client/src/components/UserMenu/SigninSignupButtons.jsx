import React from 'react';
import PropTypes from 'prop-types';
import history from '../../helpers/history';
import Button from '../util/Button';

const SigninSignupButtons = ({ closeModal }) => {
  const handleSignin = () => {
    closeModal();
    history.push('/signin');
  };
  const handleSignup = () => {
    closeModal();
    history.push('/signup');
  };

  const btnStyle = {
    margin: '1rem',
  };

  return (
    <div>
      <div className="m1">
        <p
          className="font-weight-bold"
          style={{
            marginBottom: '1rem',
            fontSize: '1.25rem',
          }}
        >
          You are not logged in
        </p>
        <p>
          Log in or sign up to continue.
        </p>
      </div>
      <div style={{ display: 'flex', justifyContent: 'flex-end' }}>
        <Button
          className="btn btn-default"
          value="Sign in"
          onClick={handleSignin}
          style={btnStyle}
        />
        <Button
          className="btn btn-default"
          value="Sign up"
          onClick={handleSignup}
          style={btnStyle}
        />
      </div>
    </div>);
};

SigninSignupButtons.propTypes = {
  closeModal: PropTypes.func.isRequired,
};

export default SigninSignupButtons;
