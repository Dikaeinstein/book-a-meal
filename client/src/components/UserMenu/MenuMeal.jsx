import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { NavLink } from 'react-router-dom';
import Modal from 'react-modal';
import MealDetail from '../Meals/MealDetail';
import MealCheckout from './MealCheckout';
import SigninSignupButtons from './SigninSignupButtons';
import Button from '../util/Button';


class MenuMeal extends Component {
  constructor(props) {
    super(props);
    this.state = {
      isMealDetailOpen: false,
      isMealChoiceOpen: false,
      meal: props.meal,
    };
    this.handleOpenMealDetailModal =
      this.handleOpenMealDetailModal.bind(this);
    this.handleCloseMealDetailModal =
      this.handleCloseMealDetailModal.bind(this);
    this.handleOpenMealChoiceModal =
      this.handleOpenMealChoiceModal.bind(this);
    this.handleCloseMealChoiceModal =
      this.handleCloseMealChoiceModal.bind(this);
  }

  handleOpenMealDetailModal() {
    this.setState({ isMealDetailOpen: true });
  }

  handleCloseMealDetailModal() {
    this.setState({ isMealDetailOpen: false });
  }

  handleOpenMealChoiceModal() {
    this.setState({ isMealChoiceOpen: true });
  }

  handleCloseMealChoiceModal() {
    this.setState({ isMealChoiceOpen: false });
  }

  render() {
    const modalStyle = {
      overlay: {
        backgroundColor: 'rgba(0, 0, 0, 0.25)',
      },
      content: {
        top: '50%',
        left: '50%',
        right: 'auto',
        bottom: 'auto',
        marginRight: '-50%',
        transform: 'translate(-50%, -50%)',
        minWidth: '20rem',
        width: '70%',
      },
    };
    const mealChoiceModalStyle = {
      ...modalStyle,
      content: { ...modalStyle.content, width: '50%' },
    };
    return (
      <div key={this.state.meal.id} className="card user-menu-item">
        <div
          className="overlay-container"
          onClick={this.handleOpenMealDetailModal}
          onKeyPress={this.handleOpenMealDetailModal}
          role="button"
          tabIndex="0"
        >
          <img
            src={this.state.meal.imageUrl}
            alt={this.state.meal.name}
          />
          <div className="overlay">
            <div className="text">
              <p>View description</p>
            </div>
          </div>
        </div>
        <Modal
          isOpen={this.state.isMealDetailOpen}
          contentLabel="Meal Detail"
          style={modalStyle}
          closeTimeoutMS={150}
        >
          <Button
            value="&times;"
            onClick={this.handleCloseMealDetailModal}
            className="close"
          />
          <MealDetail
            closeMealDetail={this.handleCloseMealDetailModal}
            meal={this.state.meal}
          />
        </Modal>
        <p className="text-black">{this.state.meal.name}</p>
        <p className="text-black">&#x20a6;{this.state.meal.price}</p>
        <Modal
          isOpen={this.state.isMealChoiceOpen}
          contentLabel="Meal Choice"
          style={mealChoiceModalStyle}
          closeTimeoutMS={150}
        >
          <Button
            value="&times;"
            onClick={this.handleCloseMealChoiceModal}
            className="close"
          />
          {this.props.loggedIn
            ?
              <MealCheckout
                meal={this.state.meal}
                closeModal={this.handleCloseMealChoiceModal}
              />
            :
              <SigninSignupButtons
                closeModal={this.handleCloseMealChoiceModal}
              />
          }
        </Modal>
        <Button
          className="btn btn-default"
          onClick={this.handleOpenMealChoiceModal}
          value="Order Now"
          style={{ margin: '1rem auto' }}
        />
      </div>
    );
  }
}

MenuMeal.propTypes = {
  meal: PropTypes.objectOf(PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.number,
  ])).isRequired,
  loggedIn: PropTypes.bool.isRequired,
};

const mapStateToProps = state => ({
  loggedIn: state.user.loggedIn,
});


export default connect(mapStateToProps)(MenuMeal);
