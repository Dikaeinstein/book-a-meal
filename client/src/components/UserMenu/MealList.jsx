import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import MenuMeal from './MenuMeal';
import { getMenu } from '../../reducers/menuReducer';


export const MealList = ({ meals, link }) => {
  const mealList = meals
    .map(meal => <MenuMeal key={meal.id} defaultMeal={meal} link={link} />);

  return (
    <section className="card user-menu text-left">
      {mealList}
    </section>
  );
};

MealList.propTypes = {
  meals: PropTypes.arrayOf(PropTypes.object).isRequired,
  link: PropTypes.string.isRequired,
};

const mapStateToProps = state => ({
  meals: getMenu(state.menu).meals,
});

export default connect(mapStateToProps)(MealList);
