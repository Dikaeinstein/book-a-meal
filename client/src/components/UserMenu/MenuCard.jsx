import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { Preloader } from 'react-materialize';
import { fetchMenu } from '../../actions/menuActions';
import MealList from './MealList';
import Loading from '../util/Loading';

class ConnectedMenuCard extends Component {
  componentDidMount() {
    this.props.fetchMenu();
  }

  render() {
    const { isFetching, link } = this.props;
    if (isFetching) {
      return (
        <div className="loader-container">
          <Loading text="Loading...">
            <Preloader flashing size="big" />
          </Loading>
        </div>
      );
    }

    return (
      <section className="main" id="menu">
        <h2 className="text-center">
          Menu for Today {(new Date()).toDateString()}
        </h2>
        <MealList link={link} />
      </section>
    );
  }
}

ConnectedMenuCard.propTypes = {
  isFetching: PropTypes.bool.isRequired,
  fetchMenu: PropTypes.func.isRequired,
  link: PropTypes.string.isRequired,
};

const mapStateToProps = state => ({
  isFetching: state.menu.isFetching,
});

const mapDispatchToProps = dispatch => ({
  fetchMenu() { dispatch(fetchMenu()); },
});

const MenuCard = connect(mapStateToProps, mapDispatchToProps)(ConnectedMenuCard);

export default MenuCard;
