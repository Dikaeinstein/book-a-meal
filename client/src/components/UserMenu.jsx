import React from 'react';
import MenuCard from './MenuCard';

const UserMenu = () => <MenuCard link="/order-confirmation" />;

export default UserMenu;
