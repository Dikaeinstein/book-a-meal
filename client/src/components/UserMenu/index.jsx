import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import MenuCard from './MenuCard';
import Footer from '../util/Footer';
import errorHandler from '../util/errorHandler';

const ConnectedUserMenu = ({ fetchMenuError }) => {
  const MenuCardWithErrorHandling =
    errorHandler(MenuCard, 'Error fetching menu');

  return (
    <div>
      <div className="bg-dark-light landing-main">
        <main style={{ minHeight: 'calc(100vh - 151px)' }}>
          <MenuCardWithErrorHandling
            link="/order-confirmation"
            error={fetchMenuError}
          />
        </main>
      </div>
      <Footer />
    </div>
  );
};

ConnectedUserMenu.propTypes = {
  /* eslint react/require-default-props: 0 */
  fetchMenuError: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.objectOf(PropTypes.string),
  ]),
};

const mapStateToProps = state => ({
  fetchMenuError: state.menus.fetchError,
});

const UserMenu = connect(mapStateToProps)(ConnectedUserMenu);

export default UserMenu;
