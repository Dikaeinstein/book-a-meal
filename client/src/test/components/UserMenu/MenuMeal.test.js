import React from 'react';
import { shallow } from 'enzyme';
import { MenuMeal } from '../../../components/UserMenu/MenuMeal';

describe('<MenuMeal />', () => {
  let meal;
  let wrapper;
  beforeAll(() => {
    meal = {
      id: 1,
      name: 'test meal',
      price: 2000,
      user: { name: 'Dika Test' },
      imageUrl: 'http://testurl.com/image.jpg',
    };

    wrapper = shallow(<MenuMeal defaultMeal={meal} />);
  });
  it('should call handleOpenMealDetail', () => {
    const spyOn = jest.spyOn(MenuMeal.prototype, 'handleOpenMealDetailModal');
    wrapper = shallow(<MenuMeal defaultMeal={meal} />);
    const overlayContainer = wrapper.find('.overlay-container');
    overlayContainer.simulate('click');
    expect(spyOn).toHaveBeenCalled();
  });
  it('should match snapshot', () => {
    const tree = shallow(<MenuMeal defaultMeal={meal} />);
    expect(tree).toMatchSnapshot();
  });
});
