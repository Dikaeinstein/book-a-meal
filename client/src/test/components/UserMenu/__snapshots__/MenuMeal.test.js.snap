// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<MenuMeal /> should match snapshot 1`] = `
<div
  className="card user-menu-item"
  key="1"
>
  <div
    className="overlay-container"
    onClick={[Function]}
    onKeyPress={[Function]}
    role="button"
    tabIndex="0"
  >
    <img
      alt="test meal"
      src="http://testurl.com/image.jpg"
    />
    <div
      className="overlay"
    >
      <div
        className="text"
      >
        <p>
          View description
        </p>
      </div>
    </div>
  </div>
  <Modal
    ariaHideApp={true}
    bodyOpenClassName="ReactModal__Body--open"
    closeTimeoutMS={150}
    contentLabel="Meal Detail"
    isOpen={false}
    parentSelector={[Function]}
    portalClassName="ReactModalPortal"
    shouldCloseOnEsc={true}
    shouldCloseOnOverlayClick={true}
    shouldFocusAfterRender={true}
    shouldReturnFocusAfterClose={true}
    style={
      Object {
        "content": Object {
          "bottom": "auto",
          "left": "50%",
          "marginRight": "-50%",
          "minWidth": "20rem",
          "right": "auto",
          "top": "50%",
          "transform": "translate(-50%, -50%)",
          "width": "70%",
        },
        "overlay": Object {
          "backgroundColor": "rgba(0, 0, 0, 0.25)",
        },
      }
    }
  >
    <Button
      className="close"
      onClick={[Function]}
      value="×"
    />
    <MealDetail
      closeMealDetail={[Function]}
      meal={
        Object {
          "id": 1,
          "imageUrl": "http://testurl.com/image.jpg",
          "name": "test meal",
          "price": 2000,
          "user": Object {
            "name": "Dika Test",
          },
        }
      }
    />
  </Modal>
  <div
    className="font-weight-bold text-black"
    style={
      Object {
        "padding": ".875rem",
      }
    }
  >
    <p>
      test meal
    </p>
    <p
      style={
        Object {
          "fontSize": "1.25rem",
        }
      }
    >
      ₦
      2000
    </p>
  </div>
  <Modal
    ariaHideApp={true}
    bodyOpenClassName="ReactModal__Body--open"
    closeTimeoutMS={150}
    contentLabel="Meal Choice"
    isOpen={false}
    parentSelector={[Function]}
    portalClassName="ReactModalPortal"
    shouldCloseOnEsc={true}
    shouldCloseOnOverlayClick={true}
    shouldFocusAfterRender={true}
    shouldReturnFocusAfterClose={true}
    style={
      Object {
        "content": Object {
          "bottom": "auto",
          "left": "50%",
          "marginRight": "-50%",
          "minWidth": "20rem",
          "right": "auto",
          "top": "50%",
          "transform": "translate(-50%, -50%)",
          "width": "45%",
        },
        "overlay": Object {
          "backgroundColor": "rgba(0, 0, 0, 0.25)",
        },
      }
    }
  >
    <Button
      className="close"
      onClick={[Function]}
      value="×"
    />
    <SigninSignupButtons
      closeModal={[Function]}
    />
  </Modal>
  <Button
    className="btn btn-default"
    onClick={[Function]}
    style={
      Object {
        "margin": "1rem auto",
      }
    }
    value="Order Now"
  />
</div>
`;
